### DISCUSSION TITLE
Reset a history state?

### DISCUSSION THREAD

**Discussion by DerekRies** [statelyai/xstate#4906] [Q&A]
This seemed like a simple question but I couldn't find the answer anywhere, so apologies if it's already posted somewhere. 

Anyway, I'm trying to reset a history state back to its initial value after some condition is reached. Is there any action/method I could call to do this?

**What I'm trying to do**

Basically I've got this mildly complicated reconnecting websocket. A caller may "close" the connection, but internally we have to drain a buffer of messages, and send a couple cleanup messages to the server before actually closing the connection. This process itself can be disrupted (read: connection lost), so that was the impetus for using a history state to return to the `connected.closing` state. However, when the machine actually does reach the `closed` state I want to reset this history state, or else subsequent attempts to "re-open" will transition to this `connected.readyToClose` state, which is really not what we want.

```ts
createMachine({
  id: "test",
  initial: "idle",
  context: {
    shouldClose: false,
  },
  states: {
    idle: {
      on: {
        REQUEST_OPEN: {
          target: "#test.connecting",
        },
      },
    },
    connecting: {
      on: {
        CONNECTION_OPENED: {
          target: "#test.connected.hist",
        },
      },
    },
    connected: {
      initial: "sending",
      states: {
        sending: {
          on: {
            REQUEST_CLOSE: {
              target: "#test.connected.closing",
            },
            SOCKET_CLOSED: {
              target: "#test.connecting",
            },
          },
        },
        closing: {
          on: {
            BUFFER_EMPTY: {
              target: "#test.connected.readyToClose",
            },
            SOCKET_CLOSED: {
              target: "#test.connecting",
            },
          },
        },
        readyToClose: {
          on: {
            CLOSE: {
              target: "#test.closed",
            },
          },
        },
        hist: {
          type: "history",
          history: "shallow",
        },
      },
    },
    closed: {
      entry: [({ context, self, system }) => {
        // I want to just reset the connected.hist history state here
      }],
    },
  },
});
```

Hopefully there's a simple way to just reset this history state back to it's initial target. Otherwise I _COULD_ take the following approaches:

 - Store something like `isClosing: true` in the context, and use a guard to transition **instead** of the history node. I'll probably end up doing this, but it felt gross _(the connected child states are fairly locked down at this point, but if that changes, there's a chance this logic will have to change as well)_ .
 - Mark `closed` as a "final" state and stop the actor, then create a new actor. This _could_ work, but then I'd have to make sure everything gets re-instantiated and re-bound as necessary. Seems like overkill.

**Comment by davidkpiano** [comment]
One potential solution could be to have "connecting after socket closed" be its own state:

<img width="617" alt="CleanShot 2024-05-20 at 16 09 36@2x" src="https://github.com/statelyai/xstate/assets/1093738/07554310-f8ff-42b6-930c-a17cc1ea5b73">

Here's the code for that:

```ts
import { setup } from "xstate";

export const machine = setup({
  types: {
    context: {} as { shouldClose: boolean },
    events: {} as
      | { type: "REQUEST_OPEN" }
      | { type: "CONNECTION_OPENED" }
      | { type: "REQUEST_CLOSE" }
      | { type: "SOCKET_CLOSED" }
      | { type: "BUFFER_EMPTY" }
      | { type: "CLOSE" },
  },
}).createMachine({
  context: {
    shouldClose: false,
  },
  id: "test",
  initial: "idle",
  states: {
    idle: {
      on: {
        REQUEST_OPEN: {
          target: "connecting",
        },
      },
    },
    connecting: {
      on: {
        CONNECTION_OPENED: {
          target: "connected",
        },
      },
    },
    connected: {
      initial: "sending",
      states: {
        sending: {
          on: {
            REQUEST_CLOSE: {
              target: "closing",
            },
            SOCKET_CLOSED: {
              target: "#test.connectingAfterSocketClosed",
            },
          },
        },
        closing: {
          on: {
            BUFFER_EMPTY: {
              target: "readyToClose",
            },
            SOCKET_CLOSED: {
              target: "#test.connectingAfterSocketClosed",
            },
          },
        },
        readyToClose: {
          on: {
            CLOSE: {
              target: "#test.closed",
            },
          },
        },
        hist: {
          type: "history",
          history: "shallow",
        },
      },
    },
    closed: {
      entry: ({ context, self, system }) => {
        // I want to just reset the connected.hist history state here
      },
    },
    connectingAfterSocketClosed: {
      on: {
        CONNECTION_OPENED: {
          target: "#test.connected.hist",
        },
      },
    },
  },
});
```
